Compiler error on flutter run causes tool to crash

1. `flutter create compiler_crash`
2. Replace lib/main.dart with:
```
import 'package:flutter/material.dart';

void main() => runApp(test: null);
```
3. In the compiler_crash/ directory, run `flutter run`.

Output:
```
$ flutter run
Launching lib/main.dart on Pixel 3 XL in debug mode...
Running Gradle task 'assembleDebug'...                          


Compiler message:

lib/main.dart:3:22: Error: Too few positional arguments: 1 required, 0 given.

void main() => runApp(test: null);

                     ^

../flutter/packages/flutter/lib/src/widgets/binding.dart:867:6: Context: Found this candidate, but the arguments don't match.

void runApp(Widget app) {

     ^^^^^^

Target kernel_snapshot failed: Exception: Errors during snapshot creation: null

#0      KernelSnapshot.build (package:flutter_tools/src/build_system/targets/dart.dart:231:7)

<asynchronous suspension>

#1      _BuildInstance._invokeInternal (package:flutter_tools/src/build_system/build_system.dart:526:25)

<asynchronous suspension>

#2      _BuildInstance.invokeTarget.<anonymous closure> (package:flutter_tools/src/build_system/build_system.dart:481:35)

#3      new Future.sync (dart:async/future.dart:222:31)

#4      AsyncMemoizer.runOnce (package:async/src/async_memoizer.dart:43:45)

#5      _BuildInstance.invokeTarget (package:flutter_tools/src/build_system/build_system.dart:481:21)

<asynchronous suspension>

<asynchronous suspension>

#6      BuildSystem.build (package:flutter_tools/src/build_system/build_system.dart:419:36)

<asynchronous suspension>

#7      AssembleCommand.runCommand (package:flutter_tools/src/commands/assemble.dart:154:50)

<asynchronous suspension>

#8      FlutterCommand.verifyThenRunCommand (package:flutter_tools/src/runner/flutter_command.dart:607:18)

<asynchronous suspension>

#9      FlutterCommand.run.<anonymous closure> (package:flutter_tools/src/runner/flutter_command.dart:507:33)

<asynchronous suspension>

#10     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:146:29)

<asynchronous suspension>

#11     _rootRun (dart:async/zone.dart:1124:13)

#12     _CustomZone.run (dart:async/zone.dart:1021:19)

#13     _runZoned (dart:async/zone.dart:1516:10)

#14     runZoned (dart:async/zone.dart:1463:12)

#15     AppContext.run (package:flutter_tools/src/base/context.dart:145:18)

<asynchronous suspension>

#16     FlutterCommand.run (package:flutter_tools/src/runner/flutter_command.dart:497:20)

#17     CommandRunner.runCommand (package:args/command_runner.dart:197:27)

<asynchronous suspension>

#18     FlutterCommandRunner.runCommand.<anonymous closure> (package:flutter_tools/src/runner/flutter_command_runner.dart:416:21)

<asynchronous suspension>

#19     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:146:29)

<asynchronous suspension>

#20     _rootRun (dart:async/zone.dart:1124:13)

#21     _CustomZone.run (dart:async/zone.dart:1021:19)

#22     _runZoned (dart:async/zone.dart:1516:10)

#23     runZoned (dart:async/zone.dart:1463:12)

#24     AppContext.run (package:flutter_tools/src/base/context.dart:145:18)

<asynchronous suspension>

#25     FlutterCommandRunner.runCommand (package:flutter_tools/src/runner/flutter_command_runner.dart:367:19)

<asynchronous suspension>

#26     CommandRunner.run.<anonymous closure> (package:args/command_runner.dart:112:25)

#27     new Future.sync (dart:async/future.dart:222:31)

#28     CommandRunner.run (package:args/command_runner.dart:112:14)

#29     FlutterCommandRunner.run (package:flutter_tools/src/runner/flutter_command_runner.dart:251:18)

#30     run.<anonymous closure>.<anonymous closure> (package:flutter_tools/runner.dart:63:22)

<asynchronous suspension>

#31     _rootRun (dart:async/zone.dart:1124:13)

#32     _CustomZone.run (dart:async/zone.dart:1021:19)

#33     _runZoned (dart:async/zone.dart:1516:10)

#34     runZoned (dart:async/zone.dart:1500:12)

#35     run.<anonymous closure> (package:flutter_tools/runner.dart:61:18)

<asynchronous suspension>

#36     AppContext.run.<anonymous closure> (package:flutter_tools/src/base/context.dart:146:29)

<asynchronous suspension>

#37     _rootRun (dart:async/zone.dart:1124:13)

#38     _CustomZone.run (dart:async/zone.dart:1021:19)

#39     _runZoned (dart:async/zone.dart:1516:10)

#40     runZoned (dart:async/zone.dart:1463:12)

#41     AppContext.run (package:flutter_tools/src/base/context.dart:145:18)

<asynchronous suspension>

#42     runInContext (package:flutter_tools/src/context_runner.dart:65:24)

<asynchronous suspension>

#43     run (package:flutter_tools/runner.dart:50:10)

#44     main (package:flutter_tools/executable.dart:65:9)

<asynchronous suspension>

#45     main (file:///usr/local/google/home/ianh/dev/flutter/packages/flutter_tools/bin/flutter_tools.dart:8:3)

#46     _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:303:32)

#47     _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:172:12)



build failed.



FAILURE: Build failed with an exception.



* Where:

Script '/usr/local/google/home/ianh/dev/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 716



* What went wrong:

Execution failed for task ':app:compileFlutterBuildDebugArm64'.

> Process 'command '/usr/local/google/home/ianh/dev/flutter/bin/flutter'' finished with non-zero exit value 1



* Try:

Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.



* Get more help at https://help.gradle.org



BUILD FAILED in 10s
Running Gradle task 'assembleDebug'...                             11.2s
Gradle task assembleDebug failed with exit code 1
```

--- a/packages/flutter_tools/lib/src/aot.dart
+++ b/packages/flutter_tools/lib/src/aot.dart
@@ -226,8 +226,11 @@ class AotBuilder {
     status?.stop();
     if (!result.success) {
       for (ExceptionMeasurement measurement in result.exceptions.values) {
-        printError(measurement.exception.toString());
-        printError(measurement.stackTrace.toString());
+        printError('Target ${measurement.target} failed: ${measurement.exception}',
+          stackTrace: measurement.fatal
+            ? measurement.stackTrace
+            : null,
+        );
       }
       throwToolExit('Failed to build aot.');
     }

--- a/packages/flutter_tools/lib/src/build_system/build_system.dart
+++ b/packages/flutter_tools/lib/src/build_system/build_system.dart
@@ -556,6 +556,8 @@ class _BuildInstance {
         }
       }
     } catch (exception, stackTrace) {
+      // TODO(jonahwilliams): throw specific exception for expected errors to mark
+      // as non-fatal. All others should be fatal.
       node.target.clearStamp(environment);
       passed = false;
       skipped = false;
@@ -573,11 +575,14 @@ class _BuildInstance {
 
 /// Helper class to collect exceptions.
 class ExceptionMeasurement {
-  ExceptionMeasurement(this.target, this.exception, this.stackTrace);
+  ExceptionMeasurement(this.target, this.exception, this.stackTrace, {this.fatal = false});
 
   final String target;
   final dynamic exception;
   final StackTrace stackTrace;
+
+  /// Whether this exception was a fatal build system error.
+  final bool fatal;
 
   @override
   String toString() => 'target: $target\nexception:$exception\n$stackTrace';

--- a/packages/flutter_tools/lib/src/bundle.dart
+++ b/packages/flutter_tools/lib/src/bundle.dart
@@ -126,8 +126,11 @@ Future<void> buildWithAssemble({
 
   if (!result.success) {
     for (ExceptionMeasurement measurement in result.exceptions.values) {
-      printError(measurement.exception.toString());
-      printError(measurement.stackTrace.toString());
+        printError('Target ${measurement.target} failed: ${measurement.exception}',
+          stackTrace: measurement.fatal
+            ? measurement.stackTrace
+            : null,
+        );
     }
     throwToolExit('Failed to build bundle.');
   }

--- a/packages/flutter_tools/lib/src/commands/assemble.dart
+++ b/packages/flutter_tools/lib/src/commands/assemble.dart
@@ -156,7 +156,7 @@ class AssembleCommand extends FlutterCommand {
     ));
     if (!result.success) {
       for (MapEntry<String, ExceptionMeasurement> data in result.exceptions.entries) {
-        printError('Target ${data.key} failed: ${data.value.exception}', stackTrace: data.value.stackTrace);
+        printError('Target ${data.key} failed: ${data.value.exception}');
       }
       throwToolExit('build failed.');
     }

--- a/packages/flutter_tools/lib/src/commands/assemble.dart
+++ b/packages/flutter_tools/lib/src/commands/assemble.dart
@@ -155,8 +155,12 @@ class AssembleCommand extends FlutterCommand {
       resourcePoolSize: argResults['resource-pool-size'],
     ));
     if (!result.success) {
-      for (MapEntry<String, ExceptionMeasurement> data in result.exceptions.entries) {
-        printError('Target ${data.key} failed: ${data.value.exception}');
+      for (ExceptionMeasurement measurement in result.exceptions.values) {
+        printError('Target ${measurement.target} failed: ${measurement.exception}',
+          stackTrace: measurement.fatal
+            ? measurement.stackTrace
+            : null,
+        );
       }
       throwToolExit('build failed.');
     }

--- a/packages/flutter_tools/lib/src/web/compile.dart
+++ b/packages/flutter_tools/lib/src/web/compile.dart
@@ -54,8 +54,11 @@ Future<void> buildWeb(
     ));
     if (!result.success) {
       for (ExceptionMeasurement measurement in result.exceptions.values) {
-        printError(measurement.stackTrace.toString());
-        printError(measurement.exception.toString());
+        printError('Target ${measurement.target} failed: ${measurement.exception}',
+          stackTrace: measurement.fatal
+            ? measurement.stackTrace
+            : null,
+        );
       }
       throwToolExit('Failed to compile application for the Web.');
     }

--- a/packages/flutter_tools/test/commands.shard/hermetic/assemble_test.dart
+++ b/packages/flutter_tools/test/commands.shard/hermetic/assemble_test.dart
@@ -54,6 +54,22 @@ void main() {
     expect(commandRunner.run(<String>['assemble', '-o Output', 'undefined']), throwsA(isInstanceOf<ToolExit>()));
   });
 
+  testbed.test('Does not log stack traces during build failure', () async {
+    final BufferLogger bufferLogger = logger;
+    final StackTrace testStackTrace = StackTrace.current;
+    when(buildSystem.build(any, any, buildSystemConfig: anyNamed('buildSystemConfig')))
+      .thenAnswer((Invocation invocation) async {
+        return BuildResult(success: false, exceptions: <String, ExceptionMeasurement>{
+          'hello': ExceptionMeasurement('hello', 'bar', testStackTrace),
+        });
+      });
+    final CommandRunner<void> commandRunner = createTestCommandRunner(AssembleCommand());
+
+    await expectLater(commandRunner.run(<String>['assemble', '-o Output', 'debug_macos_bundle_flutter_assets']), throwsA(isInstanceOf<ToolExit>()));
+    expect(bufferLogger.errorText, contains('bar'));
+    expect(bufferLogger.errorText, isNot(contains(testStackTrace.toString())));
+  });
+
   testbed.test('Only writes input and output files when the values change', () async {
     when(buildSystem.build(any, any, buildSystemConfig: anyNamed('buildSystemConfig')))
       .thenAnswer((Invocation invocation) async {

--- a/packages/flutter_tools/test/commands.shard/hermetic/assemble_test.dart
+++ b/packages/flutter_tools/test/commands.shard/hermetic/assemble_test.dart
@@ -41,7 +41,8 @@ void main() {
       });
     final CommandRunner<void> commandRunner = createTestCommandRunner(AssembleCommand());
 
-    expect(commandRunner.run(<String>['assemble', 'debug_macos_bundle_flutter_assets']), throwsA(isInstanceOf<ToolExit>()));
+    expect(commandRunner.run(<String>['assemble', 'debug_macos_bundle_flutter_assets']),
+      throwsA(isInstanceOf<ToolExit>()));
   });
 
   testbed.test('Throws ToolExit if called with non-existent rule', () async {
@@ -51,7 +52,8 @@ void main() {
       });
     final CommandRunner<void> commandRunner = createTestCommandRunner(AssembleCommand());
 
-    expect(commandRunner.run(<String>['assemble', '-o Output', 'undefined']), throwsA(isInstanceOf<ToolExit>()));
+    expect(commandRunner.run(<String>['assemble', '-o Output', 'undefined']),
+      throwsA(isInstanceOf<ToolExit>()));
   });
 
   testbed.test('Does not log stack traces during build failure', () async {
@@ -65,7 +67,8 @@ void main() {
       });
     final CommandRunner<void> commandRunner = createTestCommandRunner(AssembleCommand());
 
-    await expectLater(commandRunner.run(<String>['assemble', '-o Output', 'debug_macos_bundle_flutter_assets']), throwsA(isInstanceOf<ToolExit>()));
+    await expectLater(commandRunner.run(<String>['assemble', '-o Output', 'debug_macos_bundle_flutter_assets']),
+      throwsA(isInstanceOf<ToolExit>()));
     expect(bufferLogger.errorText, contains('bar'));
     expect(bufferLogger.errorText, isNot(contains(testStackTrace.toString())));
   });
@@ -81,7 +84,13 @@ void main() {
       });
 
     final CommandRunner<void> commandRunner = createTestCommandRunner(AssembleCommand());
-    await commandRunner.run(<String>['assemble', '-o Output', '--build-outputs=outputs', '--build-inputs=inputs', 'debug_macos_bundle_flutter_assets']);
+    await commandRunner.run(<String>[
+      'assemble',
+      '-o Output',
+      '--build-outputs=outputs',
+      '--build-inputs=inputs',
+      'debug_macos_bundle_flutter_assets',
+    ]);
 
     final File inputs = fs.file('inputs');
     final File outputs = fs.file('outputs');
@@ -91,7 +100,13 @@ void main() {
     final DateTime theDistantPast = DateTime(1991, 8, 23);
     inputs.setLastModifiedSync(theDistantPast);
     outputs.setLastModifiedSync(theDistantPast);
-    await commandRunner.run(<String>['assemble', '-o Output', '--build-outputs=outputs', '--build-inputs=inputs', 'debug_macos_bundle_flutter_assets']);
+    await commandRunner.run(<String>[
+      'assemble',
+      '-o Output',
+      '--build-outputs=outputs',
+      '--build-inputs=inputs',
+      'debug_macos_bundle_flutter_assets',
+    ]);
 
     expect(inputs.lastModifiedSync(), theDistantPast);
     expect(outputs.lastModifiedSync(), theDistantPast);
@@ -103,7 +118,13 @@ void main() {
           inputFiles: <File>[fs.file('foo'), fs.file('fizz')..createSync()],
           outputFiles: <File>[fs.file('bar'), fs.file(fs.path.join('.dart_tool', 'fizz2'))..createSync(recursive: true)]);
       });
-    await commandRunner.run(<String>['assemble', '-o Output', '--build-outputs=outputs', '--build-inputs=inputs', 'debug_macos_bundle_flutter_assets']);
+    await commandRunner.run(<String>[
+      'assemble',
+      '-o Output',
+      '--build-outputs=outputs',
+      '--build-inputs=inputs',
+      'debug_macos_bundle_flutter_assets',
+    ]);
 
     expect(inputs.readAsStringSync(), contains('foo'));
     expect(inputs.readAsStringSync(), contains('fizz'));

